README.md
# homebridge-livebox-info

homebridge-livebox-info is homebridge plugin for Apple HomeKit, get and return somes Orange Livebox informations.

[![npm](https://img.shields.io/npm/v/homebridge-livebox-info.svg)](https://www.npmjs.com/package/homebridge-livebox-info) 
[![npm](https://img.shields.io/npm/dt/homebridge-livebox-info.svg)](https://www.npmjs.com/package/homebridge-livebox-info)
[![GitHub license](https://img.shields.io/github/license/ad5030/homebridge-macosx-info.svg)](https://github.com/ad5030/homebridge-livebox-info)
<!-- [![Donate](https://img.shields.io/badge/donate-paypal-yellowgreen.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=9MC83TRGACQPJ&source=url) --> 

> **Note:** **/!\\** It's a **`beta`** version !

Such as : 
  * [x] Speed Downstream *(Mbit)*
  * [x] Speed Upstream *(Mbit)*
  * [ ] Uptime
  * [ ] SoftwareVersion
  * [ ] ExternalIPAddress
  * [ ] IPv4Address
  * [ ] IPv6Address
  * [ ] Phone Number
  * [x] Phone State
  * [x] TV State
  * [x] WifiState
<!--more-->


## Exemple of .json data response file

```json  
{  
   "updateTime":"Sat 28 Sep 2019 18:25:25 CEST",
   "Downstream":"5.74",
   "Upstream":"0.94",
   "LastChange":"Sat Sep 21 14:36:22 2019",
   "SoftwareVersion":"SR40_sip-fr-3.53.6.1_7.21.3.1",
   "UpTime":"22:17:37",
   "ExternalIPAddress":"***.***.***.***",
   "IPv4Address":"192.168.1.1",
   "IPv6Address":"****:****:****:****:****:****:****",
   "IPv6DelegatedPrefix":"****:****:**:****:****",
   "directoryNumber":"+33*********",
   "PhoneState":"Up",
   "TVState":"Available",
   "WifiState":"False"
}
```

## Prerequisites
* Install [Homebrew](https://brew.sh)<span style="color:gray"> *(Homebrew installs the stuff you need that Apple didn’t)*</span>
* Install [node.js](https://nodejs.org/en/download/package-manager/#macos) on macOS
* Install [Homebridge](https://github.com/nfarina/homebridge/wiki/Install-Homebridge-on-macOS) on macOS
* Install [Homebridge Config UI X](https://github.com/oznu/homebridge-config-ui-x#readme) on macOS <span style="color:gray">*(optional)*<span>
* Install [Eve.app](https://www.evehome.com/en/eve-app) on iOS (for all availables plugin function), or it's possible to used "Home" app on macOS (only avalable for macOS Majave) or iOS (all plugin function aren't availables on this app !)
* Install [Python 3 & pip3](https://docs.brew.sh/Homebrew-and-Python) on macOS

## Installation
Used [npm](https://www.npmjs.com/package/homebridge-livebox-info) tool to install homebridge-livebox-info, and execute the command line below

```console
npm i homebridge-livebox-info
```

## Configuration
### STEP 1 : homebridge config.json file
Add this lines in config.json
```json    
"accessories": [
        {
            "accessory": "OrangeLiveboxInfo",
            "name": "Orange Livebox Info",
            "file": "/tmp/_homebridge-livebox-info.json",
            "serial": "042-04-000",
            "updateInterval": 60000
        }
    ],
```

| Parameter       | Note | Optionnal | value | 
|-----------------|------|-----------|-------|
| `accessory`     | Name of accessory|No|`LiveboxInfo`|
| `name`          | a human-readable name for your plugin|No|`Livebox Info`|
| `file`          | .json respons file|yes|default : `/tmp/_homebridge-livebox-info.json`|
| `updateInterval`| is time in ms of data update|yes|default : `null`|

>**Note:**  
**1.** The `index.js` call *`<PATH of Node Module>/homebridge-macosx-info/sh/homebridge-livebox-info.sh`* shell script. You can find this script in the repository in `/src/sh` directory  
**2.** It's possible that you can change the path of `homebridge-livebox-info.sh` in `readUptime` function on `index.js` script

```js
async function readUptime() {
    const exec = require('child_process').exec;
    var script = await exec('/usr/local/lib/node_modules/homebridge-macosx-info/src/sh/homebridge-livebox-info.sh',
        (error, stdout, stderr) => {
            if (error !== null) {
                //this.log("exec error: " + ${error});
            }
        }); 
};
```

### STEP 2 : Adapte `homebridge-livebox-info.sh` file in `src/sh` directory
* You can change path of temporary .json files : `JSON_DATA_FILE` variable

```sh
#!/bin/sh
#-------------------------------------------------------------------
#~ @(#) Name : homebridge-livebox-info.sh
#~ @(#) Desc : Persist in file the livebox sys infrmation needed by "homebridge-livebox-info" Homebridge/HomeKit plugin
#~ @(#) version : 1.0
# Auteur : adm@di-marco.net
# Date : 2019-09-28
#-------------------------------------------------------------------
# Version history
#   v1.O - Initial version - Test and work on : mac mini (late 2014) 10.13.6 (High Sierra) & livebox v4
#-------------------------------------------------------------------
#~ Usage : homebridge-livebox-info.sh
#-------------------------------------------------------------------

DIR=$(dirname $0)
TMP_DIR=/tmp
TMP_DSLRATE=_homebridge-livebox-dslrate.tmp
TMP_PHONESTATE=_homebridge-livebox-phonestate.tmp
TMP_TVSTATE=_homebridge-livebox-tvstate.tmp
TMP_WIFISTATE=_homebridge-livebox-wifistate.tmp
JSON_DATA_FILE=$TMP_DIR/_homebridge-livebox-info.json # path of .json respons file 

function livebox_mon()
{
  $DIR/lbx_sysbus.py -dslrate > $TMP_DIR/$TMP_DSLRATE
  $DIR/lbx_sysbus.py -info >> $TMP_DIR/$TMP_DSLRATE
  $DIR/lbx_sysbus.py -phonestate > $TMP_DIR/$TMP_PHONESTATE
  $DIR/lbx_sysbus.py -tvstate > $TMP_DIR/$TMP_TVSTATE
  $DIR/lbx_sysbus.py -wifistate > $TMP_DIR/$TMP_WIFISTATE
  
  _time=`date`
  
  response=null
  key=null
  value=null
  response='{"updateTime":"'${_time}'",'

  while read line; do
    read -a fields <<< $line ;
    key=${line/%: *} ; key=$(echo "$key" | perl -pe 's/^ *| *$//g') 
    value=${line#*\:} ; value=${value:1} ; value=$(echo "$value" | perl -pe 's/^ *| *$//g') 
    if [[ "${key}" == "Upstream" || "${key}" == "Downstream" ]]
    then
      value=${value% *}
    fi
    if [ "${key}" == "UpTime" ]
    then
      value=${value% (*} ; value=${value// /}
    fi
    response=${response}\"${key}\":\"${value}\",
  done < $TMP_DIR/$TMP_DSLRATE

  response=${response%?}

  phonestate=`cat $TMP_DIR/$TMP_PHONESTATE | grep "'status'" | grep -v "{'status':"`
  phonestate=${phonestate#*\'status\':} ; phonestate=${phonestate// /} ; phonestate=${phonestate//\'/} ; phonestate=${phonestate/%,}

  tvstate=`cat $TMP_DIR/$TMP_TVSTATE | grep "'IPTVStatus'"`
  tvstate=${tvstate#*\'IPTVStatus\':} ; tvstate=${tvstate// /} ; tvstate=${tvstate%\},*} ; tvstate=${tvstate//\'/} 

  wifistate=`cat $TMP_DIR/$TMP_WIFISTATE | grep "'Status'"`
  wifistate=${wifistate#*\'Status\':} ; wifistate=${wifistate// /} ; wifistate=${wifistate//\'/} ; wifistate=${wifistate//\}/}

  response=${response%?}\",\"PhoneState\":\"$phonestate\",\"TVState\":\"$tvstate\",\"WifiState\":\"$wifistate\"}
  echo $response > $JSON_DATA_FILE
}

livebox_mon
```

### STEP 3 : Install & configure `lbx_sysbus.py` file in `src/sh` directory

`lbx_sysbus.py` is a `Python 3` script that allows you to programmatically control a Livebox and explore control possibilities and other hidden information. It is an "experimental" tool.

1. The script is written in `Python 3`. It also requires `requests` which greatly simplifies HTTP requests.

```sh
$  pip3 install requests 
```

2. Now you must change/adapt the variables to connect on Orange Livebox : `URL_LIVEBOX`, `USER_LIVEBOX`, `PASSWORD_LIVEBOX` and `VERSION_LIVEBOX` 

```sh
URL_LIVEBOX = '<url of livebox interface configuration>' #example http://192.168.1.1/
USER_LIVEBOX = '<user admin>' #example admin
PASSWORD_LIVEBOX = '<user admin pwd>' #example p@ssw0rd
VERSION_LIVEBOX = '<LiveBox version>' #example lb4
```

It will also install the Graphviz engine. On OSX we can use brew. On Linux, sudo apt-get install graphviz or equivalent depending on the distribution.

### STEP 4 : restart homebridge 
Combine the two commands in a terminal to restart homebridge background process

 - `launchctl unload ~/Library/LaunchAgents/com.homebridge.server.plist`
 - `launchctl load ~/Library/LaunchAgents/com.homebridge.server.plist`

>**Note:** This commands are only avalable for macOS 

## Todo

## Known bugs

## Credits
* The original HomeKit API work was done by [KhaosT](https://twitter.com/khaost) in his [HAP-NodeJS](https://github.com/KhaosT/HAP-NodeJS) project
* [simont77 - fakegato-history](https://github.com/simont77/fakegato-history)
* [Rene Devichi - sysbus Orange Livebox 2, 3 et 4](https://github.com/rene-d/sysbus)

## Disclaimer
I'm furnishing this software "as is". I do not provide any warranty of the item whatsoever, whether express, implied, or statutory, including, but not limited to, any warranty of merchantability or fitness for a particular purpose or any warranty that the contents of the item will be error-free. The development of this module is not supported by Apple Inc. or eve. These vendors and me are not responsible for direct, indirect, incidental or consequential damages resulting from any defect, error or failure to perform.

## License
This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details


docs/CHANGELOG.md
# homebridge-livebox-info Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## v0.0.2 2019-09-29
- Minor change in package.json

## v0.0.1 2019-09-29
 - Firt version in github public repository 


index.js
//-------------------------------------------------------------------
//~ @(#) Name : index.js
//~ @(#) Desc : 
//~ @(#) version : 1.0
// Auteur : adm@di-marco.net
// Date : 2019-09-28
//-------------------------------------------------------------------
// Version history
//   v1.O - Initial version
//   test and work on : mac mini (late 2014) & livebox v4 
//-------------------------------------------------------------------
var Accessory, Service, Characteristic, UUIDGen;
var inherits = require('util').inherits;
const fs = require('fs');
const packageFile = require("./package.json");
var os = require("os");
var hostname = os.hostname();

const UNIT_MBIT='Mbit'
const UNIT_MO='Mo'
const UNIT_PERCENT='%'
const UNIT_WATT='Watt'

module.exports = function(homebridge) {
    if(!isConfig(homebridge.user.configPath(), "accessories", "OrangeLiveboxInfo")) {
        return;
    }
    
    Accessory = homebridge.platformAccessory;
    Service = homebridge.hap.Service;
    Characteristic = homebridge.hap.Characteristic;
    UUIDGen = homebridge.hap.uuid;
//    FakeGatoHistoryService = require("fakegato-history")(homebridge);

    homebridge.registerAccessory('homebridge-livebox-info', 'OrangeLiveboxInfo', OrangeLiveboxInfo);
}

async function readUptime() {
    const exec = require('child_process').exec;
var script = await exec('/usr/local/lib/node_modules/homebridge-livebox-info/src/sh/homebridge-livebox-info.sh', 
(error, stdout, stderr) => {
            if (error !== null) {
                //this.log("exec error: " + ${error});
            }
        }); 
};

function isConfig(configFile, type, name) {
    var config = JSON.parse(fs.readFileSync(configFile));
    if("accessories" === type) {
        var accessories = config.accessories;
        for(var i in accessories) {
            if(accessories[i]['accessory'] === name) {
                return true;
            }
        }
    } else if("platforms" === type) {
        var platforms = config.platforms;
        for(var i in platforms) {
            if(platforms[i]['platform'] === name) {
                return true;
            }
        }
    } else {
    }
    return false;
};

function OrangeLiveboxInfo(log, config) {
    if(null == config) {
        return;
    }

    this.log = log;
	this.name = config["name"];
    if(config["file"]) {
		this.readFile = config["file"];
    } else {
		this.readFile = "/tmp/_homebridge-livebox-info.json";
	}
	if(config["updateInterval"] && config["updateInterval"] > 0) {
        this.updateInterval = config["updateInterval"];
    } else {
        this.updateInterval = null;
    }
  
	this.setUpServices();
};

OrangeLiveboxInfo.prototype.getDownstream = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var Downstream = (obj.Downstream);
	callback(null, Downstream);
};

OrangeLiveboxInfo.prototype.getUpstream = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var Upstream = obj.Upstream;
	callback(null, Upstream);
};

OrangeLiveboxInfo.prototype.getPhoneState = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var PhoneState = obj.PhoneState;
	callback(null, PhoneState);
};

OrangeLiveboxInfo.prototype.getTVState = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var TVState = obj.TVState;
	callback(null, TVState);
};

OrangeLiveboxInfo.prototype.getWifiState = function (callback) {
	var json = fs.readFileSync(this.readFile, "utf-8");
	var obj = JSON.parse(json);
	var WifiState = obj.WifiState;
	callback(null, WifiState);
};


OrangeLiveboxInfo.prototype.setUpServices = function () {

	var that = this;
	var temp;
	
	this.infoService = new Service.AccessoryInformation();
	this.infoService
		.setCharacteristic(Characteristic.Manufacturer, "@ad5030")
		.setCharacteristic(Characteristic.Model, this.name)
		.setCharacteristic(Characteristic.SerialNumber, "042-SN-20190928" + "_" + packageFile.version)
		.setCharacteristic(Characteristic.FirmwareRevision, packageFile.version);
	
//	this.fakeGatoHistoryService = new FakeGatoHistoryService("weather", this, { storage: 'fs' });
	
	let uuid1 = UUIDGen.generate(that.name + '-Downstream');
	downstream = function (displayName, subtype) {
		Characteristic.call(this, 'Downstream', uuid1);
		this.setProps({
			format: Characteristic.Formats.STRING,
			unit: UNIT_MBIT,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(downstream, Characteristic);
	downstream.UUID = uuid1;

	let uuid2 = UUIDGen.generate(that.name + '-Upstream');
	upstream = function () {
		Characteristic.call(this, 'Upstream', uuid2);
		this.setProps({
			format: Characteristic.Formats.STRING,
			unit: UNIT_MBIT,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(upstream, Characteristic);
	upstream.UUID = uuid2;
	
	let uuid3 = UUIDGen.generate(that.name + '-PhoneState');
	phonestate = function () {
		Characteristic.call(this, 'Phone State', uuid3);
		this.setProps({
			format: Characteristic.Formats.STRING,
			//unit: UNIT_MO,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(phonestate, Characteristic);
	phonestate.UUID = uuid3;

	let uuid4 = UUIDGen.generate(that.name + '-TVState');
	tvstate = function () {
		Characteristic.call(this, 'TV State', uuid4);
		this.setProps({
			format: Characteristic.Formats.STRING,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(tvstate, Characteristic);
	tvstate.UUID = uuid4;

	let uuid5 = UUIDGen.generate(that.name + '-WifiState');
	wifistate = function () {
		Characteristic.call(this, 'Wifi State', uuid5);
		this.setProps({
			format: Characteristic.Formats.STRING,
			//unit: UNIT_PERCENT,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	inherits(wifistate, Characteristic);
	wifistate.UUID = uuid5;


	this.liveboxService = new Service.TemperatureSensor(that.name);
	var currentTemperatureCharacteristic = this.liveboxService.getCharacteristic(Characteristic.CurrentTemperature);		
	this.liveboxService.getCharacteristic(downstream)
		.on('get', this.getDownstream.bind(this));
	this.liveboxService.getCharacteristic(upstream)
		.on('get', this.getUpstream.bind(this));
	this.liveboxService.getCharacteristic(phonestate)
		.on('get', this.getPhoneState.bind(this));
	this.liveboxService.getCharacteristic(tvstate)
		.on('get', this.getTVState.bind(this));
	this.liveboxService.getCharacteristic(wifistate)
		.on('get', this.getWifiState.bind(this));
//	function getCurrentTemperature() {
//		var data = fs.readFileSync(that.readFile, "utf-8");

//		var obj = JSON.parse(data);
//		var temperatureVal = (obj.temperature);

//		temp = temperatureVal;
//		that.log.debug("update currentTemperatureCharacteristic value: " + temperatureVal);
//		return temperatureVal;
//	}
	
	readUptime();
	
//	currentTemperatureCharacteristic.updateValue(getCurrentTemperature());
//	if(that.updateInterval) {
//		setInterval(() => {
//			currentTemperatureCharacteristic.updateValue(getCurrentTemperature());
			
//			that.log("Current Temp: " + temp);
//			this.fakeGatoHistoryService.addEntry({time: new Date().getTime() / 1000, temp: temp});
			//this.fakeGatoHistoryService.addEntry({time: new Date().getTime(), temp: temp});	

//			readUptime();
			
//		}, that.updateInterval);
//	}
	
//	currentTemperatureCharacteristic.on('get', (callback) => {
//		callback(null, getCurrentTemperature());
//	});
}

OrangeLiveboxInfo.prototype.getServices = function () {

	return [this.infoService, this.liveboxService];
};

private/.sync.sh
#!/bin/sh
#-------------------------------------------------------------------
#~ @(#) Name : .sync.sh
#~ @(#) Desc : Sync projet and deploy in local host for test
#~ @(#) version : na
# Auteur : adm@di-marco.net
# Date : 2019-05-01
#-------------------------------------------------------------------
# Version history
#   na
#-------------------------------------------------------------------
#~ Usage : .sync.sh
#-------------------------------------------------------------------

ENV=PORDUCTION
SRC=../$(dirname $0)
USER_TARGET='AD' # User name
HOST_TARGET='192.168.42.3' #host for test only
PATH_TARGET='~/r2d2/it/homekit/homebridge-livebox-info' # Path of script directory 
NODE_MODULE_TARGET='/usr/local/lib/node_modules/homebridge-livebox-info'

PORT_TARGET=22 # ssh port (inside port is 22) 

PLIST_TARGET='~/Library/LaunchAgents/' # Directory for plist file 
CONFIG_HOMEBRIDGE_TARGET='~/.homebridge/'

red=$'\e[1;31m%s\e[0m\n'
gre=$'\e[1;32m%s\e[0m\n'
yel=$'\e[1;33m%s\e[0m\n'
blu=$'\e[1;34m%s\e[0m\n'
mag=$'\e[1;35m%s\e[0m\n'
l_blu=$'\e[1;36m%s\e[0m\n'
end=$'\e[0m'

clear
printf "$red" "/!\ [$ENV] /!\ "
printf "$l_blu" "src directory [$SRC]"
printf "$l_blu" "users[$USER_TARGET]"
printf "$l_blu" "Host [$HOST_TARGET]"
printf "$l_blu" "Path target [$PATH_TARGET]"
printf "$l_blu" "Config homebridge target [$CONFIG_HOMEBRIDGE_TARGET]"
printf "$l_blu" "Port target [$PORT_TARGET]"

chmod +x $SRC/src/sh/*.sh
chmod +x $SRC/src/sh/*.py

printf "$red" "-> save in $PATH_TARGET ..."
/usr/bin/ssh -p $PORT_TARGET $USER_TARGET@$HOST_TARGET "mkdir -p ${PATH_TARGET}/src/sh"
/usr/bin/ssh -p $PORT_TARGET $USER_TARGET@$HOST_TARGET "mkdir -p ${PATH_TARGET}/docs"
/usr/bin/ssh -p $PORT_TARGET $USER_TARGET@$HOST_TARGET "mkdir -p ${PATH_TARGET}/img"

scp -P $PORT_TARGET -rp $SRC/LICENSE $USER_TARGET@$HOST_TARGET:$PATH_TARGET/LICENSE
scp -P $PORT_TARGET -rp $SRC/package.json $USER_TARGET@$HOST_TARGET:$PATH_TARGET/package.json
scp -P $PORT_TARGET -rp $SRC/README.md $USER_TARGET@$HOST_TARGET:$PATH_TARGET/README.md
scp -P $PORT_TARGET -rp $SRC/index.js $USER_TARGET@$HOST_TARGET:$PATH_TARGET/index.js
scp -P $PORT_TARGET -rp $SRC/src/sh/homebridge-livebox-info.sh $USER_TARGET@$HOST_TARGET:$PATH_TARGET/src/sh/homebridge-livebox-info.sh
scp -P $PORT_TARGET -rp $SRC/src/sh/lbx_sysbus.py $USER_TARGET@$HOST_TARGET:$PATH_TARGET/src/sh/lbx_sysbus.py
scp -P $PORT_TARGET -rp $SRC/docs/CHANGELOG.md $USER_TARGET@$HOST_TARGET:$PATH_TARGET/docs/CHANGELOG.md
#scp -P $PORT_TARGET -rp $SRC/img/homebridge-livebox-info_screenshots.png $USER_TARGET@$HOST_TARGET:$PATH_TARGET/img/homebridge-livebox-info_screenshots.png

printf "$red" "-> Deploy in $NODE_MODULE_TARGET ..."
/usr/bin/ssh -p $PORT_TARGET $USER_TARGET@$HOST_TARGET "sudo mkdir -p ${NODE_MODULE_TARGET}/src/sh"
/usr/bin/ssh -p $PORT_TARGET $USER_TARGET@$HOST_TARGET "sudo mkdir -p ${NODE_MODULE_TARGET}/docs"
/usr/bin/ssh -p $PORT_TARGET $USER_TARGET@$HOST_TARGET "sudo mkdir -p ${NODE_MODULE_TARGET}/img"

/usr/bin/ssh -p $PORT_TARGET $USER_TARGET@$HOST_TARGET "sudo chown -R AD:staff ${NODE_MODULE_TARGET}"

scp -P $PORT_TARGET -rp $SRC/LICENSE $USER_TARGET@$HOST_TARGET:$NODE_MODULE_TARGET/LICENSE
scp -P $PORT_TARGET -rp $SRC/README.md $USER_TARGET@$HOST_TARGET:$NODE_MODULE_TARGET/README.md
scp -P $PORT_TARGET -rp $SRC/package.json $USER_TARGET@$HOST_TARGET:$NODE_MODULE_TARGET/package.json
scp -P $PORT_TARGET -rp $SRC/index.js $USER_TARGET@$HOST_TARGET:$NODE_MODULE_TARGET/index.js
scp -P $PORT_TARGET -rp $SRC/src/sh/homebridge-livebox-info.sh $USER_TARGET@$HOST_TARGET:$NODE_MODULE_TARGET/src/sh/homebridge-livebox-info.sh
scp -P $PORT_TARGET -rp $SRC/src/sh/lbx_sysbus.py $USER_TARGET@$HOST_TARGET:$NODE_MODULE_TARGET/src/sh/lbx_sysbus.py
scp -P $PORT_TARGET -rp $SRC/docs/CHANGELOG.md $USER_TARGET@$HOST_TARGET:$NODE_MODULE_TARGET/docs/CHANGELOG.md
#scp -P $PORT_TARGET -rp $SRC/img/homebridge-livebox-info_screenshots.png $USER_TARGET@$HOST_TARGET:$NODE_MODULE_TARGET/img/homebridge-livebox-info_screenshots.png

printf "$red" "  |_ homebridge restart"
/usr/bin/ssh -p $PORT_TARGET $USER_TARGET@$HOST_TARGET "launchctl unload ~/Library/LaunchAgents/com.homebridge.server.plist && rm ~/.homebridge/config.json" && printf "$gre" "    |_ homebride [OFF]"
/usr/bin/ssh -p $PORT_TARGET $USER_TARGET@$HOST_TARGET "ln -s ~/.homebridge/config.wake.json ~/.homebridge/config.json && launchctl load ~/Library/LaunchAgents/com.homebridge.server.plist" && printf "$gre" "    |_ homebride [ON]"

printf "$red" "-> Sync ownCloud repository ..."
/usr/bin/ssh -p $PORT_TARGET $USER_TARGET@$HOST_TARGET "/usr/local/Cellar/php@7.2/7.2.15/bin/php /Library/WebServer/Documents/owncloud/occ files:scan AD"

printf "$gre" "[Ended deployement]"

src/sh/homebridge-livebox-info.sh
#!/bin/sh
#-------------------------------------------------------------------
#~ @(#) Name : homebridge-livebox-info.sh
#~ @(#) Desc : Persist in file the livebox sys infrmation needed by "homebridge-livebox-info" Homebridge/HomeKit plugin
#~ @(#) version : 1.0
# Auteur : adm@di-marco.net
# Date : 2019-09-28
#-------------------------------------------------------------------
# Version history
#   v1.O - Initial version - Test and work on : mac mini (late 2014) 10.13.6 (High Sierra) & livebox v4
#-------------------------------------------------------------------
#~ Usage : homebridge-livebox-info.sh
#-------------------------------------------------------------------

DIR=$(dirname $0)
TMP_DIR=/tmp
TMP_DSLRATE=_homebridge-livebox-dslrate.tmp
TMP_PHONESTATE=_homebridge-livebox-phonestate.tmp
TMP_TVSTATE=_homebridge-livebox-tvstate.tmp
TMP_WIFISTATE=_homebridge-livebox-wifistate.tmp
JSON_DATA_FILE=$TMP_DIR/_homebridge-livebox-info.json # path of .json respons file 

function livebox_mon()
{
  $DIR/lbx_sysbus.py -dslrate > $TMP_DIR/$TMP_DSLRATE
  $DIR/lbx_sysbus.py -info >> $TMP_DIR/$TMP_DSLRATE
  $DIR/lbx_sysbus.py -phonestate > $TMP_DIR/$TMP_PHONESTATE
  $DIR/lbx_sysbus.py -tvstate > $TMP_DIR/$TMP_TVSTATE
  $DIR/lbx_sysbus.py -wifistate > $TMP_DIR/$TMP_WIFISTATE
  
  _time=`date`
  
  response=null
  key=null
  value=null
  response='{"updateTime":"'${_time}'",'

  while read line; do
    read -a fields <<< $line ;
    key=${line/%: *} ; key=$(echo "$key" | perl -pe 's/^ *| *$//g') 
    value=${line#*\:} ; value=${value:1} ; value=$(echo "$value" | perl -pe 's/^ *| *$//g') 
    if [[ "${key}" == "Upstream" || "${key}" == "Downstream" ]]
    then
      value=${value% *}
    fi
    if [ "${key}" == "UpTime" ]
    then
      value=${value% (*} ; value=${value// /}
    fi
    response=${response}\"${key}\":\"${value}\",
  done < $TMP_DIR/$TMP_DSLRATE
  
  response=${response%?}

  phonestate=`cat $TMP_DIR/$TMP_PHONESTATE | grep "'status'" | grep -v "{'status':"`
  phonestate=${phonestate#*\'status\':} ; phonestate=${phonestate// /} ; phonestate=${phonestate//\'/} ; phonestate=${phonestate/%,}

  tvstate=`cat $TMP_DIR/$TMP_TVSTATE | grep "'IPTVStatus'"`
  tvstate=${tvstate#*\'IPTVStatus\':} ; tvstate=${tvstate// /} ; tvstate=${tvstate%\},*} ; tvstate=${tvstate//\'/} 

  wifistate=`cat $TMP_DIR/$TMP_WIFISTATE | grep "'Status'"`
  wifistate=${wifistate#*\'Status\':} ; wifistate=${wifistate// /} ; wifistate=${wifistate//\'/} ; wifistate=${wifistate//\}/}

  response=${response%?}\",\"PhoneState\":\"$phonestate\",\"TVState\":\"$tvstate\",\"WifiState\":\"$wifistate\"}
  echo $response > $JSON_DATA_FILE
}

livebox_mon